src/stream/stream.rs

```diff
@@ -118,6 +120,39 @@ pub trait Stream {
         }
     }
 
+    /// Returns the element that gives the minimum value with respect to the
+    /// specified comparison function. If several elements are equally minimum,
+    /// the first element is returned. If the stream is empty, `None` is returned.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # fn main() { async_std::task::block_on(async {
+    /// #
+    /// use std::collections::VecDeque;
+    /// use async_std::stream::Stream;
+    ///
+    /// let s: VecDeque<usize> = vec![1, 2, 3].into_iter().collect();
+    ///
+    /// let min = Stream::min_by(s.clone(), |x, y| x.cmp(y)).await;
+    /// assert_eq!(min, Some(1));
+    ///
+    /// let min = Stream::min_by(s, |x, y| y.cmp(x)).await;
+    /// assert_eq!(min, Some(3));
+    ///
+    /// let min = Stream::min_by(VecDeque::<usize>::new(), |x, y| x.cmp(y)).await;
+    /// assert_eq!(min, None);
+    /// #
+    /// # }) }
+    /// ```
+    fn min_by<F>(self, compare: F) -> MinBy<Self, F>
```

Comment:
The return type here should use `ret!` like in `fn any`, for example. This function would ideally be an `async fn` if Rust supported that syntax in traits today.

Reply:
How is `ret!` supposed to work for
1. Futures that take `self` instead of `&mut self` and thus should have no lifetime parameters
2. Streams
?